LPS-76620

jorgediaz-lr opened the pull request:
https://issues.liferay.com/browse/LPS-76620

In order to solve LPS-76620, there are two pulls: https://github.com/brunobasto/liferay-portal/pull/891 (portal repo) and https://github.com/liferay/com-liferay-dynamic-data-mapping/pull/589 (this)

Please, review both pulls

liferay-continuous-integration commented:
Comment ci:test to trigger CI tests.

liferay-continuous-integration commented:
<h1>Build started.</h1><p>Jenkins is currently running tests.</p><h4>Base Branch:</h4><p>Branch Name: <a href="https://github.com/liferay/com-liferay-dynamic-data-mapping/tree/master">master</a></p><h4>Job Summary:</h4><p>Job Link: <a href="https://test-1-1.liferay.com/job/test-subrepository-acceptance-pullrequest(master)/285/">test-subrepository-acceptance-pullrequest(master)</a></p><p><b>For more details click <a href="https://test-1-1.liferay.com//userContent/jobs/test-subrepository-acceptance-pullrequest(master)/builds/285/jenkins-report.html">here</a>.</b></p>

liferay-continuous-integration commented:
<h1>Validation FAILED.</h1><p>Build Time: 17 minutes 56 seconds</p><h4>Base Branch:</h4><p>Branch Name: <a href="https://github.com/liferay/com-liferay-dynamic-data-mapping/tree/master">master</a><br />Branch GIT ID: <a href="https://github.com/liferay/com-liferay-dynamic-data-mapping/commit/43c4bb54a220705530ada88b863233586ca4c059">43c4bb54a220705530ada88b863233586ca4c059</a></p><h4>Task Summary:</h4><ul><li><strong><a href="https://test-1-1.liferay.com//userContent/jobs/test-subrepository-acceptance-pullrequest(master)/builds/285/subrepository-compile-jdk8.log">subrepository-compile-jdk8</a></strong> -  :white_check_mark:</li><li><strong><a href="https://test-1-1.liferay.com//userContent/jobs/test-subrepository-acceptance-pullrequest(master)/builds/285/subrepository-semantic-versioning-jdk8.log">subrepository-semantic-versioning-jdk8</a></strong> -  :white_check_mark:</li><li><strong><a href="https://test-1-1.liferay.com//userContent/jobs/test-subrepository-acceptance-pullrequest(master)/builds/285/subrepository-pmd-jdk8.log">subrepository-pmd-jdk8</a></strong> -  :white_check_mark:</li><li><strong><a href="https://test-1-1.liferay.com//userContent/jobs/test-subrepository-acceptance-pullrequest(master)/builds/285/subrepository-source-format-jdk8.log">subrepository-source-format-jdk8</a></strong> -  :white_check_mark:</li><li><strong><a href="https://test-1-1.liferay.com//userContent/jobs/test-subrepository-acceptance-pullrequest(master)/builds/285/subrepository-unit-jdk8.log">subrepository-unit-jdk8</a></strong> -  :x:<h6>Job Results:</h6><p>556 Tests Passed.<br />9 Tests Failed.</p><ol><li><a href="https://test-1-7.liferay.com/job/test-subrepository-acceptance-pullrequest-validation(master)/143//testReport/com.liferay.dynamic.data.mapping.form.web.internal.notification/DDLFormEmailNotificationSenderTest/testGetField">DDLFormEmailNotificationSenderTest.testGetField</a></li><li><a href="https://test-1-7.liferay.com/job/test-subrepository-acceptance-pullrequest-validation(master)/143//testReport/com.liferay.dynamic.data.mapping.form.web.internal.notification/DDLFormEmailNotificationSenderTest/testGetFieldWithNullValue">DDLFormEmailNotificationSenderTest.testGetFieldWithNullValue</a></li><li><a href="https://test-1-7.liferay.com/job/test-subrepository-acceptance-pullrequest-validation(master)/143//testReport/com.liferay.dynamic.data.mapping.io.internal/DDMFormXSDDeserializerTest/testDDMFormAndFieldsDefaultLocales">DDMFormXSDDeserializerTest.testDDMFormAndFieldsDefaultLocales</a></li><li>...</li></ol><p><strong>Click <a href="https://test-1-7.liferay.com/job/test-subrepository-acceptance-pullrequest-validation(master)/143//testReport/">here</a> for more failures.</strong></p></li></ul><h5>For more details click <a href="https://test-1-7.liferay.com/job/test-subrepository-acceptance-pullrequest-validation(master)/143/">here</a>.</h5>

jorgediaz-lr commented:
ci:retest

liferay-continuous-integration commented:
Pull request test invoked at http://test-1-6.liferay.com/job/test-subrepository-acceptance-pullrequest(master).

liferay-continuous-integration commented:
<h1>Build started.</h1><p>Jenkins is currently running tests.</p><h4>Base Branch:</h4><p>Branch Name: <a href="https://github.com/liferay/com-liferay-dynamic-data-mapping/tree/master">master</a></p><h4>Job Summary:</h4><p>Job Link: <a href="https://test-1-6.liferay.com/job/test-subrepository-acceptance-pullrequest(master)/197/">test-subrepository-acceptance-pullrequest(master)</a></p><p><b>For more details click <a href="https://test-1-6.liferay.com//userContent/jobs/test-subrepository-acceptance-pullrequest(master)/builds/197/jenkins-report.html">here</a>.</b></p>

liferay-continuous-integration commented:
<h1>Validation FAILED.</h1><p>Build Time: 17 minutes 30 seconds</p><h4>Base Branch:</h4><p>Branch Name: <a href="https://github.com/liferay/com-liferay-dynamic-data-mapping/tree/master">master</a><br />Branch GIT ID: <a href="https://github.com/liferay/com-liferay-dynamic-data-mapping/commit/3c4ad135ed0880808ab62c563b3fed13885bb4f7">3c4ad135ed0880808ab62c563b3fed13885bb4f7</a></p><h4>Task Summary:</h4><ul><li><strong><a href="https://test-1-6.liferay.com//userContent/jobs/test-subrepository-acceptance-pullrequest(master)/builds/197/subrepository-compile-jdk8.log">subrepository-compile-jdk8</a></strong> -  :white_check_mark:</li><li><strong><a href="https://test-1-6.liferay.com//userContent/jobs/test-subrepository-acceptance-pullrequest(master)/builds/197/subrepository-semantic-versioning-jdk8.log">subrepository-semantic-versioning-jdk8</a></strong> -  :white_check_mark:</li><li><strong><a href="https://test-1-6.liferay.com//userContent/jobs/test-subrepository-acceptance-pullrequest(master)/builds/197/subrepository-pmd-jdk8.log">subrepository-pmd-jdk8</a></strong> -  :white_check_mark:</li><li><strong><a href="https://test-1-6.liferay.com//userContent/jobs/test-subrepository-acceptance-pullrequest(master)/builds/197/subrepository-source-format-jdk8.log">subrepository-source-format-jdk8</a></strong> -  :white_check_mark:</li><li><strong><a href="https://test-1-6.liferay.com//userContent/jobs/test-subrepository-acceptance-pullrequest(master)/builds/197/subrepository-unit-jdk8.log">subrepository-unit-jdk8</a></strong> -  :x:<h6>Job Results:</h6><p>321 Tests Passed.<br />2 Tests Failed.</p><ol><li><strong><a href="https://test-1-17.liferay.com/job/test-subrepository-acceptance-pullrequest-validation(master)/108/">test-subrepository-acceptance-pullrequest-validation&#40;master&#41; #108</a></strong><pre><code>     [echo] 	at org.gradle.launcher.GradleMain.main(GradleMain.java:23)
     [echo] 	at org.gradle.wrapper.BootstrapMainStarter.start(BootstrapMainStarter.java:31)
     [echo] 	at org.gradle.wrapper.WrapperExecutor.execute(WrapperExecutor.java:108)
     [echo] 	at org.gradle.wrapper.GradleWrapperMain.main(GradleWrapperMain.java:61)
     [echo] Caused by: org.gradle.api.internal.tasks.compile.CompilationFailedException: Compilation failed; see the compiler error output for details.
     [echo] 	at org.gradle.api.internal.tasks.compile.JdkJavaCompiler.execute(JdkJavaCompiler.java:48)
     [echo] 	at org.gradle.api.internal.tasks.compile.JdkJavaCompiler.execute(JdkJavaCompiler.java:33)
     [echo] 	at org.gradle.api.internal.tasks.compile.NormalizingJavaCompiler.delegateAndHandleErrors(NormalizingJavaCompiler.java:104)
     [echo] 	at org.gradle.api.internal.tasks.compile.NormalizingJavaCompiler.execute(NormalizingJavaCompiler.java:53)
     [echo] 	at org.gradle.api.internal.tasks.compile.NormalizingJavaCompiler.execute(NormalizingJavaCompiler.java:38)
     [echo] 	at org.gradle.api.internal.tasks.compile.CleaningJavaCompilerSupport.execute(CleaningJavaCompilerSupport.java:35)
     [echo] 	at org.gradle.api.internal.tasks.compile.CleaningJavaCompilerSupport.execute(CleaningJavaCompilerSupport.java:25)
     [echo] 	at org.gradle.api.tasks.compile.JavaCompile.performCompilation(JavaCompile.java:206)
     [echo] 	at org.gradle.api.tasks.compile.JavaCompile.compile(JavaCompile.java:187)
     [echo] 	at org.gradle.api.tasks.compile.JavaCompile.compile(JavaCompile.java:130)
     [echo] 	at org.gradle.internal.reflect.JavaMethod.invoke(JavaMethod.java:73)
     [echo] 	at org.gradle.api.internal.project.taskfactory.DefaultTaskClassInfoStore$IncrementalTaskAction.doExecute(DefaultTaskClassInfoStore.java:163)
     [echo] 	at org.gradle.api.internal.project.taskfactory.DefaultTaskClassInfoStore$StandardTaskAction.execute(DefaultTaskClassInfoStore.java:134)
     [echo] 	at org.gradle.api.internal.project.taskfactory.DefaultTaskClassInfoStore$StandardTaskAction.execute(DefaultTaskClassInfoStore.java:123)
     [echo] 	at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.executeAction(ExecuteActionsTaskExecuter.java:95)
     [echo] 	at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.executeActions(ExecuteActionsTaskExecuter.java:76)
     [echo] 	... 62 more
     [echo] </code></pre></li></ol></li></ul><h5>For more details click <a href="https://test-1-17.liferay.com/job/test-subrepository-acceptance-pullrequest-validation(master)/108/">here</a>.</h5>

leoadb commented:
ci:retest

liferay-continuous-integration commented:
Pull request test invoked at http://test-1-1.liferay.com/job/test-subrepository-acceptance-pullrequest(master).

liferay-continuous-integration commented:
<h1>Build started.</h1><p>Jenkins is currently running tests.</p><h4>Base Branch:</h4><p>Branch Name: <a href="https://github.com/liferay/com-liferay-dynamic-data-mapping/tree/master">master</a></p><h4>Job Summary:</h4><p>Job Link: <a href="https://test-1-1.liferay.com/job/test-subrepository-acceptance-pullrequest(master)/292/">test-subrepository-acceptance-pullrequest(master)</a></p><p><b>For more details click <a href="https://test-1-1.liferay.com//userContent/jobs/test-subrepository-acceptance-pullrequest(master)/builds/292/jenkins-report.html">here</a>.</b></p>

liferay-continuous-integration commented:
<h1>Validation FAILED.</h1><p>Build Time: 22 minutes 14 seconds</p><h4>Base Branch:</h4><p>Branch Name: <a href="https://github.com/liferay/com-liferay-dynamic-data-mapping/tree/master">master</a><br />Branch GIT ID: <a href="https://github.com/liferay/com-liferay-dynamic-data-mapping/commit/2f05e302fe2899bd0d90d04fea727dc71f18c6d8">2f05e302fe2899bd0d90d04fea727dc71f18c6d8</a></p><h4>Task Summary:</h4><ul><li><strong><a href="https://test-1-1.liferay.com//userContent/jobs/test-subrepository-acceptance-pullrequest(master)/builds/292/subrepository-compile-jdk8.log">subrepository-compile-jdk8</a></strong> -  :white_check_mark:</li><li><strong><a href="https://test-1-1.liferay.com//userContent/jobs/test-subrepository-acceptance-pullrequest(master)/builds/292/subrepository-semantic-versioning-jdk8.log">subrepository-semantic-versioning-jdk8</a></strong> -  :white_check_mark:</li><li><strong><a href="https://test-1-1.liferay.com//userContent/jobs/test-subrepository-acceptance-pullrequest(master)/builds/292/subrepository-pmd-jdk8.log">subrepository-pmd-jdk8</a></strong> -  :white_check_mark:</li><li><strong><a href="https://test-1-1.liferay.com//userContent/jobs/test-subrepository-acceptance-pullrequest(master)/builds/292/subrepository-source-format-jdk8.log">subrepository-source-format-jdk8</a></strong> -  :white_check_mark:</li><li><strong><a href="https://test-1-1.liferay.com//userContent/jobs/test-subrepository-acceptance-pullrequest(master)/builds/292/subrepository-unit-jdk8.log">subrepository-unit-jdk8</a></strong> -  :x:<h6>Job Results:</h6><p>321 Tests Passed.<br />2 Tests Failed.</p><ol><li><strong><a href="https://test-1-4.liferay.com/job/test-subrepository-acceptance-pullrequest-validation(master)/183/">test-subrepository-acceptance-pullrequest-validation&#40;master&#41; #183</a></strong><pre><code>     [echo] 	at org.gradle.launcher.GradleMain.main(GradleMain.java:23)
     [echo] 	at org.gradle.wrapper.BootstrapMainStarter.start(BootstrapMainStarter.java:31)
     [echo] 	at org.gradle.wrapper.WrapperExecutor.execute(WrapperExecutor.java:108)
     [echo] 	at org.gradle.wrapper.GradleWrapperMain.main(GradleWrapperMain.java:61)
     [echo] Caused by: org.gradle.api.internal.tasks.compile.CompilationFailedException: Compilation failed; see the compiler error output for details.
     [echo] 	at org.gradle.api.internal.tasks.compile.JdkJavaCompiler.execute(JdkJavaCompiler.java:48)
     [echo] 	at org.gradle.api.internal.tasks.compile.JdkJavaCompiler.execute(JdkJavaCompiler.java:33)
     [echo] 	at org.gradle.api.internal.tasks.compile.NormalizingJavaCompiler.delegateAndHandleErrors(NormalizingJavaCompiler.java:104)
     [echo] 	at org.gradle.api.internal.tasks.compile.NormalizingJavaCompiler.execute(NormalizingJavaCompiler.java:53)
     [echo] 	at org.gradle.api.internal.tasks.compile.NormalizingJavaCompiler.execute(NormalizingJavaCompiler.java:38)
     [echo] 	at org.gradle.api.internal.tasks.compile.CleaningJavaCompilerSupport.execute(CleaningJavaCompilerSupport.java:35)
     [echo] 	at org.gradle.api.internal.tasks.compile.CleaningJavaCompilerSupport.execute(CleaningJavaCompilerSupport.java:25)
     [echo] 	at org.gradle.api.tasks.compile.JavaCompile.performCompilation(JavaCompile.java:206)
     [echo] 	at org.gradle.api.tasks.compile.JavaCompile.compile(JavaCompile.java:187)
     [echo] 	at org.gradle.api.tasks.compile.JavaCompile.compile(JavaCompile.java:130)
     [echo] 	at org.gradle.internal.reflect.JavaMethod.invoke(JavaMethod.java:73)
     [echo] 	at org.gradle.api.internal.project.taskfactory.DefaultTaskClassInfoStore$IncrementalTaskAction.doExecute(DefaultTaskClassInfoStore.java:163)
     [echo] 	at org.gradle.api.internal.project.taskfactory.DefaultTaskClassInfoStore$StandardTaskAction.execute(DefaultTaskClassInfoStore.java:134)
     [echo] 	at org.gradle.api.internal.project.taskfactory.DefaultTaskClassInfoStore$StandardTaskAction.execute(DefaultTaskClassInfoStore.java:123)
     [echo] 	at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.executeAction(ExecuteActionsTaskExecuter.java:95)
     [echo] 	at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.executeActions(ExecuteActionsTaskExecuter.java:76)
     [echo] 	... 62 more
     [echo] </code></pre></li></ol></li></ul><h5>For more details click <a href="https://test-1-4.liferay.com/job/test-subrepository-acceptance-pullrequest-validation(master)/183/">here</a>.</h5>

leoadb commented:
@jorgediaz-lr Real errors here. Can you please resend? Thanks.

jorgediaz-lr commented:
ci:retest

liferay-continuous-integration commented:
Pull request test invoked at http://test-1-15.liferay.com/job/test-subrepository-acceptance-pullrequest(master).

liferay-continuous-integration commented:
<h1>Build started.</h1><p>Jenkins is currently running tests.</p><h4>Base Branch:</h4><p>Branch Name: <a href="https://github.com/liferay/com-liferay-dynamic-data-mapping/tree/master">master</a></p><h4>Job Summary:</h4><p>Job Link: <a href="https://test-1-15.liferay.com/job/test-subrepository-acceptance-pullrequest(master)/91/">test-subrepository-acceptance-pullrequest(master)</a></p><p><b>For more details click <a href="https://test-1-15.liferay.com//userContent/jobs/test-subrepository-acceptance-pullrequest(master)/builds/91/jenkins-report.html">here</a>.</b></p>

liferay-continuous-integration commented:
<h1>Validation FAILED.</h1><p>Build Time: 24 minutes 28 seconds</p><h4>Base Branch:</h4><p>Branch Name: <a href="https://github.com/liferay/com-liferay-dynamic-data-mapping/tree/master">master</a><br />Branch GIT ID: <a href="https://github.com/liferay/com-liferay-dynamic-data-mapping/commit/52fd8e53c99d18b4f2c4c9be388defa7ed406ac8">52fd8e53c99d18b4f2c4c9be388defa7ed406ac8</a></p><h4>Task Summary:</h4><ul><li><strong><a href="https://test-1-15.liferay.com//userContent/jobs/test-subrepository-acceptance-pullrequest(master)/builds/91/subrepository-compile-jdk8.log">subrepository-compile-jdk8</a></strong> -  :white_check_mark:</li><li><strong><a href="https://test-1-15.liferay.com//userContent/jobs/test-subrepository-acceptance-pullrequest(master)/builds/91/subrepository-semantic-versioning-jdk8.log">subrepository-semantic-versioning-jdk8</a></strong> -  :white_check_mark:</li><li><strong><a href="https://test-1-15.liferay.com//userContent/jobs/test-subrepository-acceptance-pullrequest(master)/builds/91/subrepository-pmd-jdk8.log">subrepository-pmd-jdk8</a></strong> -  :white_check_mark:</li><li><strong><a href="https://test-1-15.liferay.com//userContent/jobs/test-subrepository-acceptance-pullrequest(master)/builds/91/subrepository-source-format-jdk8.log">subrepository-source-format-jdk8</a></strong> -  :white_check_mark:</li><li><strong><a href="https://test-1-15.liferay.com//userContent/jobs/test-subrepository-acceptance-pullrequest(master)/builds/91/subrepository-unit-jdk8.log">subrepository-unit-jdk8</a></strong> -  :x:<h6>Job Results:</h6><p>321 Tests Passed.<br />2 Tests Failed.</p><ol><li><strong><a href="https://test-1-2.liferay.com/job/test-subrepository-acceptance-pullrequest-validation(master)/226/">test-subrepository-acceptance-pullrequest-validation&#40;master&#41; #226</a></strong><pre><code>     [echo] 	at org.gradle.launcher.GradleMain.main(GradleMain.java:23)
     [echo] 	at org.gradle.wrapper.BootstrapMainStarter.start(BootstrapMainStarter.java:31)
     [echo] 	at org.gradle.wrapper.WrapperExecutor.execute(WrapperExecutor.java:108)
     [echo] 	at org.gradle.wrapper.GradleWrapperMain.main(GradleWrapperMain.java:61)
     [echo] Caused by: org.gradle.api.internal.tasks.compile.CompilationFailedException: Compilation failed; see the compiler error output for details.
     [echo] 	at org.gradle.api.internal.tasks.compile.JdkJavaCompiler.execute(JdkJavaCompiler.java:48)
     [echo] 	at org.gradle.api.internal.tasks.compile.JdkJavaCompiler.execute(JdkJavaCompiler.java:33)
     [echo] 	at org.gradle.api.internal.tasks.compile.NormalizingJavaCompiler.delegateAndHandleErrors(NormalizingJavaCompiler.java:104)
     [echo] 	at org.gradle.api.internal.tasks.compile.NormalizingJavaCompiler.execute(NormalizingJavaCompiler.java:53)
     [echo] 	at org.gradle.api.internal.tasks.compile.NormalizingJavaCompiler.execute(NormalizingJavaCompiler.java:38)
     [echo] 	at org.gradle.api.internal.tasks.compile.CleaningJavaCompilerSupport.execute(CleaningJavaCompilerSupport.java:35)
     [echo] 	at org.gradle.api.internal.tasks.compile.CleaningJavaCompilerSupport.execute(CleaningJavaCompilerSupport.java:25)
     [echo] 	at org.gradle.api.tasks.compile.JavaCompile.performCompilation(JavaCompile.java:206)
     [echo] 	at org.gradle.api.tasks.compile.JavaCompile.compile(JavaCompile.java:187)
     [echo] 	at org.gradle.api.tasks.compile.JavaCompile.compile(JavaCompile.java:130)
     [echo] 	at org.gradle.internal.reflect.JavaMethod.invoke(JavaMethod.java:73)
     [echo] 	at org.gradle.api.internal.project.taskfactory.DefaultTaskClassInfoStore$IncrementalTaskAction.doExecute(DefaultTaskClassInfoStore.java:163)
     [echo] 	at org.gradle.api.internal.project.taskfactory.DefaultTaskClassInfoStore$StandardTaskAction.execute(DefaultTaskClassInfoStore.java:134)
     [echo] 	at org.gradle.api.internal.project.taskfactory.DefaultTaskClassInfoStore$StandardTaskAction.execute(DefaultTaskClassInfoStore.java:123)
     [echo] 	at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.executeAction(ExecuteActionsTaskExecuter.java:95)
     [echo] 	at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.executeActions(ExecuteActionsTaskExecuter.java:76)
     [echo] 	... 62 more
     [echo] </code></pre></li></ol></li></ul><h5>For more details click <a href="https://test-1-2.liferay.com/job/test-subrepository-acceptance-pullrequest-validation(master)/226/">here</a>.</h5>

jorgediaz-lr commented:
@leoadb previous error was a missing dependence that I have added in new commit https://github.com/liferay/com-liferay-dynamic-data-mapping/pull/589/commits/b48c5a3b77a2df6c08059d00a8d415d7983bc541

Now there is a false positive in DDMFormEmailNotificationSenderTest because I changed a String with a SoyHTMLContextValue object
```
java.lang.AssertionError: expected:<test> but was:<com.liferay.portal.template.soy.utils.SoyHTMLContextValue@3cd842ba>
	at org.junit.Assert.fail(Assert.java:88)
	at org.junit.Assert.failNotEquals(Assert.java:834)
	at org.junit.Assert.assertEquals(Assert.java:118)
	at org.junit.Assert.assertEquals(Assert.java:144)
	at com.liferay.dynamic.data.mapping.form.web.internal.notification.DDMFormEmailNotificationSenderTest.testGetField(DDMFormEmailNotificationSenderTest.java:73)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
```

I will have a look to that error tomorrow

jorgediaz-lr commented:
After speaking with @jbalsas we have to do additional work in SoyTemplate side (related pull request https://github.com/brunobasto/liferay-portal/pull/891) 

I am closing both pr and I will send them again with the new solution

pedroqueiroz94 commented:
Hey @jorgediaz-lr , thanks for the heads up! 

Whenever you're ready, please send the changes related to ddl-form-web directly to our subrepo (**com-liferay-dynamic-data-mapping/dynamic-data-mapping-form-web**), instead of sending to the central repo. 

You can see further information in this loop post: https://loop.liferay.com/home/-/loop/feed/9102278 .

Thank you for your comprehension,

Happy new year!
